### DISCUSSION TITLE
v45.0-alpha1

### DISCUSSION THREAD

**Discussion by westnordost** [streetcomplete/StreetComplete#4113] [Release Discussions]
<h3>🍰 Overlays!</h3>
  <p>Make every survey unique! How? Dedicate each to updating a different aspect of the map by activating a different overlay, respectively!</p>
  <p>This is <b>for advanced users</b> who want to be more efficient in adding and maintaining certain data area-wide. Read more in <a href="https://www.openstreetmap.org/user/westnordost/diary/399378">this blog post</a>!<br/>For now, there are only overlays for <i>street lighting</i> and for <i>sidewalks</i>, more will be added over time. 🎉 (#2461, #2506, #3920)</p>
<p float="left">
<img  src="https://user-images.githubusercontent.com/4661658/173657351-c51b63c8-a7f8-448a-b675-d85d6a75a5cf.png"> <img   src="https://user-images.githubusercontent.com/4661658/173657357-ebb4966e-73a4-4d41-bb5f-f2073142f0eb.png">
</p>
  <h3>New Quests</h3>
  <ul>
  <li><i>What kind of memorial is this?</i> (#4102, #1572), by @matkoniecz, thanks @kmpoppe, @naposm, @FloEdelmann</li>
  </ul>
  <h3>Enhancements and Fixes</h3>
  <ul>
  <li>Other small enhancements (#3919, ...)</li>
  </ul>

<hr /><em>This discussion was created from the release <a href='https://github.com/streetcomplete/StreetComplete/releases/tag/v45.0-alpha1'>v45.0-alpha1</a>.</em>

**Comment by jacobwhall** [comment]
I can't wait to give overlays a try. It was interesting to read about how local data analysis sped up the time to download quests. Thank you Tobias!

**Comment by mcliquid** [comment]
I have ether problem that the "more option" and the last selection in the bottom left corner are overlying each other. Is this a purpose?

![Screenshot_20220615-070751](https://user-images.githubusercontent.com/3351668/173740981-b2226e4b-1298-40d0-aaca-e7447eb4a139.png)

**Reply by westnordost** [reply]
Not really. I could either move this button up some bit.

Or, I could re-think how the whole overlay form is displayed. E.g. instead of the floating "..." and the floating "OK" button, there could be a button bar with these to items. However, that would waste some space.

**Reply by mcliquid** [reply]
@westnordost From my side it would be enough if you move the three dots menu icon to the right side maybe?

**Reply by FloEdelmann** [reply]
Or move the <kbd>⋮</kbd> button to the top right? Although it would lose some of its similarity with the "More answers…" button in quest forms.

**Reply by westnordost** [reply]
Well in another overlay, there could just as well be an element at the position the "..." would be moved to.

**Reply by westnordost** [reply]
For now I'll just move the last answer button up a bit

**Comment by rugk** [comment]
Awesome updtae and looks like a great feature! I canÄt wait to try it out!

However as for [your quest guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/OVERLAY_GUIDELINES.md), I donÄt get one point:
> **Not a quest:** Usually eligible data should either be recorded and displayed in an overlay or as a quest.

Are not your examples/implementations contradictions of that? They are _both_ a quest _and_ an overlay, i.e. the lit and sidewalk quests can _also_ be answered as quests, too?
I also don't see any reason to make them mutually exclusive, how one wants to record the data mostly depends on whether the user wants to map all (normal mode) or focus on something specific (overlay).

**Reply by westnordost** [reply]
Yes that is true, from that point of view, both current overlays are against my own guidelines. Here are my reasons:

- the overlay guidelines are deliberately written maybe more strict as need-be as a form of self-protection: I don't want this issue tracker to be swamped with requests to create an overlay for this and that (, and for bench backrests too 😉). At least, the first overlays that should make it into the app should also be ones that are really meaningful, not cover some fringe interest

- the "street lighting" overlay exists because it is often very helpful to see how the neighbouring roads and paths have been tagged when tagging `lit=*`. Also, it is easy to spot "mistakes" with this, because after all, the tag is a little fuzzy (because indirect light), one user may have answered that there is no light at one section of footway, another that yes,... it's a little messy. So the overlay creates the oppurtunity to equalize this a bit

- the sidewalk overlay exists among other reasons because the quest is limited in that it doesn't even ask when there is *some* parallel footway near because it *could* be a sidewalk. You know, because SC is very cautious in asking non-advanced users for too many elements. This leaves holes in the data which can be observed and filled with this overlay, especially helpful on large intersections

- quests work best to supply information on isolated features that have no or little relation to surrounding data. It is trivial to answer whether a bus stop has a shelter. It is much less easy and quick to answer anything about features that span longer distances (any road or path) because you need to check the whole highlighted stretch, sometimes even beyond (maxspeed quest) or ideally look at how surrounding data is tagged (lit quest, smoothness quest). 
This means, by the way, that often to solve one of those quests, you need to go past a couple of other quests first and then finally you hopefully still remember what the smoothness or whatever was at the start of the street.
So, since quests have been the only thing there is for a long time in this app, everything has been made into quests. You could maybe say, shoehorned into quests. Especially in consideration of [this statistic](https://youtu.be/sta5e0KqS3s?t=460) and recurring feedback we get on this issue tracker (last #3981), maybe none but the dead easy to answer quests should be quests and the rest be reserved for advanced users? Sure, you **can** have a quest for street lighting etc., but if there were good reasons for it to become an overlay, then the overlay is maybe the more fitting format. In my opinion, there is little reason to allow recording data in two different ways, one of which is less efficient and more prone to error because of e.g. missing context

So, this particular rule should also start a discussion, or at least the thought process/notion that maybe, `lit`, `sidewalk` and anything that gets added as overlays later should ideally actually not be a quest at the same time. Or if it is, there should be a good reason for it.

**Reply by westnordost** [reply]
So, the current quests that I think should at least be considered to turn into overlays and *remove the quest* are:
- lit
- smoothness (currently disabled by default)
- maxspeed (currently disabled by default)
- sidewalk surface (currently disabled by default)
- sidewalk
- cycleway
- lanes
- parking lanes
- housenumbers

**Reply by matkoniecz** [reply]
> housenumbers

definitely no for "remove the quest"  part: quest works perfectly fine and is quite rare in general. I would not keep enabling layer to check is there a missing housenumber nearby, while I added some every month while editing with SC

> lanes, cycleway, sidewalk, maxspeed, smoothness, sidewalk surface

for this quest seems to work fine, I see no benefit from "remove the quest" part. Though layer may be useful ESPECIALLY for maxspeed. Which is unfortunately nasty to implement (`maxspeed:forward` / `maxspeed:backward` waves hello)

**Reply by mnalis** [reply]
While I have yet to dedicate more time to fully evaluate Overlays (I've just gave it a cursory try), I'm also **_strongly against removing the quests_** just because they _"become available as layers"_. 

For example, when I map "with my bicycle hat" only, I want to map `cycleway` **and** `sidewalk` **and** `surface` **and** `smoothness` **and** `lit`. Using quest preset with just those quests works fine. If, however, they were _removed_ as quests (just because someone implemented the Overlay for them), I'd have to either:

- do 5 times the amount of surveying trips, one time with each Overlay
- every few dozen meters change the active Overlay to cycle through all 5 of them

Both solutions sound like they would be hugely annoying and massive reduction in usability, especially in weakly mapped areas.

So **please do not remove the quests** just because the similar overlay becomes available. Perhaps (if something must be done), matching quest might be _temporarily disabled while its equivalent overlay is enabled_ (i.e. if I enable `lit overlay`, SC might temporarily not show `lit quests`; but start showing them again as soon as overlays were disabled or different overlay chosen).

**Reply by westnordost** [reply]
> So please do not remove the quests just because the similar overlay becomes available. Perhaps (if something must be done), matching quest might be temporarily disabled while its equivalent overlay is enabled (i.e. if I enable lit overlay, SC might temporarily not show lit quests; but start showing them again as soon as overlays were disabled or different overlay chosen).

I was thinking about implementing that. I gather it makes sense to implement this, but only if it is determined that we definitely do not want to go with the "either overlay or quest" solution. So, I want to first put this option to discussion.

**Reply by matkoniecz** [reply]
Though for `lit` and `smoothness` where consistency in data (at least local) is much more useful I would be fine with quest removal.

**Reply by mnalis** [reply]
@matkoniecz for me it would still suck even if only such a subset of quests I'm interested in were removed :cry:. Perhaps in your area the map is mostly solved so you'd use **overlay mostly for verification-only where it would be noticeably less problematic**; but in my case it is mostly in areas with very little (if any) detail.

So _if they were removed as quests_, I'd still have to solve `lit` overlay, then switch to `smoothness`, then solve that, then switch back to `lit`, then solve that, then switch to `smoothness` etc. - and doing it every dozen seconds ad infinitum would be horrible user experience (I can speak only for myself, of course; although I can't see how it would not be a serious downgrade for anybody in same situation - i.e. being **interested in solving multiple overlays in areas where they are mostly unsolved**). 

Especially when I'm not on foot with lots of leisure time (where it would still be very annoying and a waste of time compared to current situation, but at least doable), but instead in a moving vehicle (like a bus/coach), which would practically guarantee I'd need to make several trips to solve quests which are important to me (like `lit` and `smoothness`). The problem gets even more compounded when the way is split into many smaller segments (as it happens often here), because there is then a very limited time to answer any quest/overlay.

**Reply by matkoniecz** [reply]
@mnalis no, in my area `lit` is also mostly not done - but here value of solving quests in isolation from existing data is much lower. 

But I also see why you would prefer to keep it.

**Reply by mnalis** [reply]
And while I agree that consistency of data is important, I find it _much less problematic_ if some segment of the road is mapped `smoothness=good` while most others are mapped as `smoothness=excellent` (e.g. due to different mappers subjectivity), then if there were **no** `smoothness` data **at all** because I couldn't enter it because the functional `smoothness quest` was removed and overlay concept was too troublesome / time consuming to use (due to need to constantly switch between different overlays).

Also, if quests were removed when similar overlay was implemented; (and we noted that overlays are directed more at advanced users), doesn't that mean that **majority** (i.e. not-advanced users) would then be **deprived of solving those important things** (to them) at all? Wouldn't then every new overlay that was added actually push StreetComplete more and more to advanced-users-only tool, as more and more functionality would not be available to majority of its (current) target demographics (i.e. non-advanced "quest-only" users)?

I think users who want simplicity should be allowed to solve the quests in the simple way as they did before; even if they would miss some context and thus perhaps sometimes give less-then-perfect answer. The more advanced users could then come after them and tweak/unify it (kind of like [TaskingManager](https://wiki.openstreetmap.org/wiki/Tasking_Manager)  mapping/verifying - even if nobody ever verifies, the mapping part is hugely important. Of course it would be better if verified, but...). Simplicity is what made SC popular; and while I'm all for expert-enabled SC edition, I do not think it should come at the expense of common ("simple") users of SC.

**Reply by rugk** [reply]
> I find it much less problematic if some segment of the road is mapped smoothness=good while most others are mapped as smoothness=excellent

I agree here. Tend to think it from the data consumer's view, I guess having 90% correct data (`smoothness=excellent`) and having 10% data that varies a bit (`smoothness=good`) is likely not a big problem, given your routing algorithm is anyway limited (in what time or so it estimates, due to many factors) and this minor mistakes will not make such a big impact.
Having no data at all may have an impact though as only an average can be used.
But that's just my guess.

In addition to what has been said, I guess I would be fine with _disabling_ quests when you opt-in into overlays. Given overlays are for advanced users only maybe they could be disabled by default (though promoted e.g. with an achievement or a button on the home page which enables it), then:
* user enables overlays -> gets the question (or just information?) that/whether the corresponding quests would/should be disabled
* this automatically disable the corresponding quests
* users could re-enable them in the settings

I agree `housenumbers` should always be kept as a quest as it is rare and quite worthwhile to survey.

---

Also, maybe in general a way to _quickly switch overlays_ (including the "default view") on the home of SC is needed, because I could imagine myself wanting to check _each_ view/overlay, just to see whether there is something red and it thus makes sense to map/look into that.

**Comment by matkoniecz** [comment]
I an testing layer functionality and I must say that it is amazing and really useful!

**Comment by matkoniecz** [comment]
@westnordost Is overlay functionality stable enough to get PR/commits? Or are you still tweaking it heavily (I plan to work on the most important of improvement ideas recently posted).

Also, are you planning to add surface overlay? Or rejected it for some reason? Or prefer to avoid adding new layers for now? (surface layer seems quite promising to me, for multiple reasons)

**Reply by westnordost** [reply]
It's stable as far as I am concerned. Today I went on a 600-quest-long survey and tested it heavily. I found one small issue and fixed it.

**Reply by westnordost** [reply]
I was thinking about a smoothness overlay. It would heavily profit from this feature as the smoothness is heavily dependent on how the surrounding roads are tagged for comparison - both to have consistent tagging and as a decision aid which is the most fitting. These two are so important IMO (after shortly testing out the smoothness quest) that the smoothness quest is the first candidate for complete removal in favour of an overlay that does the same, in fact.

Now, the smoothness quest, as currently implemented, is only asked for a very limited number of surfaces (asphalt, concrete, compacted, paving stones), because we only have descriptions + images for that. Plus, a smoothness can only be defined **after** the surface has been defined. It would be a little disruptive to only show anything for that overlay if the surface is defined and is one of these values.

This is why I thought maybe there could be a surface+smoothness overlay. So there is one button at the top to select the surface, and if one of asphalt, concrete,... etc. is selected, another will appear below where the smoothness can be selected.
(Only) issue that the color palette would be quite big. First for smoothness (greenish to orangish), but then also for where smoothness is missing because it cannot be selected (maybe brownish colors) and then also for paving stones, asphalt etc. where no smoothness is set yet.

The value of a just-surface-layer is also not very apparent to me. It does not seem to help the surveyor find the most fitting value, the value selected is not really dependent on how things have been tagged surrounding it (unlike lit, smoothness) and the quest already shows every single way where the information is missing - unlike the sidewalk or cycleway quest, where for certain ways, it is not asked.

**Reply by matkoniecz** [reply]
> The value of a just-surface-layer is also not very apparent to me.

Mostly for verification. I keep encountering tracks tagged with completely wrong surfaces, and this is often clustered - but checking all of them in Vespucci is too time-consuming and too annoying

**Reply by westnordost** [reply]
Okay. The reason why I mention this smoothness(+surface) overlay is because it's the question whether a "just surface" overlay makes sense if there is an overlay for smoothness+surface.

**Reply by westnordost** [reply]
I figure, even if (the decision would be that) a "smoothness" overlay would also allow to tag the surface, that isn't really a barrier to have an overlay to "just" tag the surface. Users interested in mapping surfaces shouldn't be pushed to also supply the very subjective smoothness tag.

So, I think if you want to add such a layer, feel free to do so. (But maybe first finish the parking condition stuff?)

**Comment by mnalis** [comment]
Could we do something before (non-alpha) release about modified hamburger menu? As the (often-used) `Scan for quests here` is now **much harder** to click correctly - see https://github.com/streetcomplete/StreetComplete/discussions/4130.

I could probably do a PR for easier options (like [moving team mode to settings, or around, or as another icon](https://github.com/streetcomplete/StreetComplete/discussions/4130#discussioncomment-2981837)) if @westnordost would choose what would be acceptable?

**Reply by matkoniecz** [reply]
I am not sure should we wait for that (putting every single decision on @westnordost is not the best solution either).

FYI, I was planning to make "scan for quest here" much larger to avoid this problem - if noone else does something else which will work.

**Reply by westnordost** [reply]
If it is difficult to tap due to the size, those list items could have a bigger padding

**Reply by mnalis** [reply]
While "fat fingers" does indeed compound the problem, the main issue is that automatic pattern recognition of human brain fails - people (after fist few times) _**don't actually read**_ what the screen says, but brain positions finger automatically at _"wide text rectangle on bottom of white form (or below 4 squares) in the middle of the screen"_. 
Which worked before, but now there are **two** such wide rectangles, so there is 50-50 chance one would hit correct one. Which is a big problem as that is very frequently used button.

**Reply by westnordost** [reply]
Bit pressing the wrong button is not a destructive action. The direct link to the profile screen is certainly more obstructive because it opens a new activity rather than a dialog. I have accidentally clicked on it countless times

**Reply by mnalis** [reply]
What I liked about previous UI is that `Scan for quest here` was:
- it was **_only_** such shaped rectangle, **and** located at the bottom of the hamburger dialog, which made it easy for my brain to hit it automatically (without invoking slower and more power-intensive reading & decision-making higher brain functions)
- it was **full length** from left to right, which made it easily accessible to both right-handed and left-handed people (even when used one-handed).

I'd like to keep that good parts. Would you accept PR which moved `Enter team mode` rectangle **_above_** the `My Profile`/`Overlay` buttons? That would be quick fix which would keep both usability of `Scan for quests here`, as well the muscle memory for it for current users, while not negatively impacting discoverability of `team mode`?

(Maybe also making `Scan for quests here` slightly bigger (taller) while we're at it, to tolerate for "fat fingers"?)

<details>

<summary>digression</summary>

Related to activity/dialog distinction: not that it is very related to issue at hand, but I actually find misclicking on `Enter team mode` **slightly more obstructive** as a misclick requires 4 new clicks to fix (`back` to remove keyboard, `back` to remove dialog, `hamburger`, `Scan for quests here`). If I were to misclick on `My Profile` instead of `Settings` for example (which I incidentally don't, as they are each in another corner of a big square, so my brain seems to cope with that just fine without reading the texts), it would require "only" 3 clicks to fix (`back`, `hamburger`, `Settings`).

And while I agree it is not **destructive**, it **is** highly annoying - because `Scan for quests here` is **very often** used (esp. when in fast moving vehicle where it is only way to use SC), and trying to be fast only increases the error rate.

</details>

**Reply by westnordost** [reply]
> I'd like to keep that good parts. Would you accept PR which moved Enter team mode rectangle above the My Profile/Overlay buttons? 

No, that would be inconsistent.

**Reply by mnalis** [reply]
How about making the the square grid 2x3 instead of current 2x2, and putting `Team mode` back there as a square as it was (leaving one square free for future feature, as it was before initial team mode implementation)? That would be consistent with previous behaviors, while not having usability drawbacks that current  v45.0-alpha1 has? Would you accept PR for that?

**Reply by westnordost** [reply]
Well that was the state before, and it looked ugly.

**Reply by mnalis** [reply]
It worked fine, though. We could also make `Preset quick change` to fill that empty square in 2x3 grid, which has been requested several times [before](https://github.com/streetcomplete/StreetComplete/issues/3827). That would additionally both fix the ugliness, and make many power-users happy. Would you accept that?

**Reply by westnordost** [reply]
Right now, no. I am not prepared to acccept any micro-optimizations in UI regarding overlays because I want to first get users accostumed to it and then later we can review how the UI can be re-worked. Also potentially on a larger scale.

**Reply by mnalis** [reply]
I have no problem with Overlays UI being fluid while its being worked on (it is to be expected); my issue is that current hamburger form (which contains much more than Overlays - specifically very often used `Scan for quest here`) is IMHO **_serious UI usability regression_** compared to previous version (v43.2), so I just wanted to help there before it is dropped on unsuspecting users. 
I plan to anyway fix it for my fork _in some way_ as it bugs me a lot, so I'd though to offer PR here too if you'd like some of [suggestions](https://github.com/streetcomplete/StreetComplete/discussions/4130) - so that work doesn't have to be done twice. But if I read that right and you're firmly unwilling to reconsider any UI change in hamburger menu currently, I understand and will stop bugging you about that subject. :heart:

**Comment by boothym** [comment]
ITO World used to have something like this on their website where you could browse for various things like speed limits, surfaces as well - is it possible for this overlay to be used outside of StreetComplete or on its own page?

**Reply by westnordost** [reply]
I know, I remember. It was part of the inspiration for overlays, too bad it does not exist anymore. It is not possible to use StreetComplete overlays outside the app on Android (e.g. on a website). I would say the modern day replacement for ITO World maps are either of
- https://mapcomplete.osm.be/ (e.g. [street lighting theme](https://mapcomplete.osm.be/street_lighting.html) )
- https://www.mapcontrib.xyz (e.g. [sidewalks theme](https://www.mapcontrib.xyz/t/0f786a-Sidewalks) )

The former has more active development, I think.
